beluga/test/benchmark/benchmark_raycasting.cpp

```diff
@@ -50,12 +50,124 @@ BENCHMARK_CAPTURE(BM_Bresenham2i, Modified, beluga::Bresenham2i::kModified)
+    [[nodiscard]] bool is_unknown(bool) const { return false; }
+    [[nodiscard]] bool is_occupied(bool value) const { return value; }
+  };
+
+  explicit BaselineGrid(std::initializer_list<bool>, double resolution) : resolution_{resolution} {
+    std::fill(std::begin(data()), std::end(data()), false);
+  }
+
+  [[nodiscard]] const Sophus::SE2d& origin() const { return origin_; }
+
+  [[nodiscard]] auto& data() { return grid_; }
+  [[nodiscard]] const auto& data() const { return grid_; }
+  [[nodiscard]] std::size_t size() const { return grid_.size(); }
+
+  [[nodiscard]] std::size_t width() const { return Cols; }
+  [[nodiscard]] std::size_t height() const { return Rows; }
+  [[nodiscard]] double resolution() const { return resolution_; }
+
+  [[nodiscard]] auto value_traits() const { return ValueTraits{}; }
+
+ private:
+  double resolution_;
+  Sophus::SE2d origin_;
+  std::array<bool, Rows * Cols> grid_;
+  static constexpr std::size_t kWidth = Cols;
+  static constexpr std::size_t kHeight = Rows;
+};
+
+template <class Map>
+std::optional<double> baseline_raycast(const Map& map, Eigen::Vector2i source, Eigen::Vector2i target) {
+  const bool steep = std::abs(target.y() - source.y()) > std::abs(target.x() - source.x());
+
+  if (steep) {
+    std::swap(source.x(), source.y());
+    std::swap(target.x(), target.y());
+  }
+
+  const auto delta = Eigen::Vector2i{
+      std::abs(target.x() - source.x()),
+      std::abs(target.y() - source.y()),
+  };
+
+  int error = 0;
+
+  const auto step = Eigen::Vector2i{
+      source.x() < target.x() ? 1 : -1,
+      source.y() < target.y() ? 1 : -1,
+  };
+
+  auto current = source;
+
+  do {
+    if (steep) {
+      if (map.data_at(current.y(), current.x()).value_or(true)) {
+        break;
+      }
+    } else {
+      if (map.data_at(current.x(), current.y()).value_or(true)) {
+        break;
+      }
+    }
+
+    if (current.x() == (target.x() + step.x())) {
+      return std::nullopt;
+    }
+
+    current.x() += step.x();
+    error += delta.y();
+    if (delta.x() <= 2 * error) {
+      current.y() += step.y();
+      error -= delta.x();
+    }
+  } while (true);
+
+  return (current - source).norm() * map.resolution();
+}
+
+template <template <std::size_t, std::size_t> class Grid>
+void BM_RayCasting2d_BaselineRaycast(benchmark::State& state) {
+  constexpr double kMaxRange = 100.0;
+  constexpr double kResolution = 0.05;
+
+  Grid<1280, 1280> map{{}, kResolution};
+
+  const auto n = static_cast<int>(state.range(0));
+  map.data()[map.index_at(n, n)] = true;
+
+  const auto source_pose = Eigen::Vector2d{1., 1.};
+  const auto beam_bearing = Sophus::SO2d{Sophus::Constants<double>::pi() / 4.};
+
+  const auto source = map.cell_near(source_pose);
+  const auto target = map.cell_near(source_pose + kMaxRange * beam_bearing.unit_complex());
+
+  for (auto _ : state) {
+    benchmark::DoNotOptimize(baseline_raycast(map, source, target));
```

Comment:
Do we know if this is returning the right result, JIC?

Reply:
Yes, after https://github.com/Ekumen-OS/beluga/pull/201/commits/f9690b72e832811ceb22dcb7b947d13456634b6d.